File: 백준/Silver/5014. 스타트링크/스타트링크.py
from collections import deque
F, S, G, U, D = map(int, input().split())
que= deque()
que.append([S,0])
visited = [False]*(F+1)
flag = False
while que:
    currNode, currDist = que.popleft()
    if currNode==G:
        print(currDist)
        flag = True
        break
    for step in [+U, -D]:
        temp = step+currNode
        if 0<temp<=F and not(visited[temp]):
            que.append([temp, currDist+1])
            visited[temp]=True

if not(flag):
    print("use the stairs")

File: 백준/Silver/5014. 스타트링크/스타트링크.js
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.on("line", (line) => {
    const [F, S, G, U, D] = line.split(" ").map(Number);
    
    function isValid(step) {
        return step >= 1 && step <= F;
    }
    
    function bfs() {
        const queue = [[S, 0]];
        const visited = Array(F + 1).fill(false);
        visited[S] = true;
        
        while (queue.length > 0) {
            const [now, process] = queue.shift();
            
            if (now === G) return process;
            
            if (isValid(now + U) && !visited[now + U]) {
                queue.push([now + U, process + 1]);
                visited[now + U] = true;
            }
            if (isValid(now - D) && !visited[now - D]) {
                queue.push([now - D, process + 1]);
                visited[now - D] = true;
            }
        }
        return "use the stairs";
    }
    
    console.log(bfs());
    rl.close();
});
